### PR TITLE
Mesh position and index buffers are now required to exist

### PR DESCRIPTION
**Pull request by VinhTruongAmbient**
The rest of the code base asserts their existence so it makes sense to enforce it at import-time.

### CODE REVIEW

File: crates/model_import/src/gltf/mod.rs [hunk 1]
```diff
@@ -52,22 +51,9 @@ pub async fn import(import: &GltfImport, asset_crate: &mut ModelCrate) -> anyhow
...
-                                let mut chunk = chunk.collect_vec();
-                                chunk.swap(1, 2);
-                                chunk
-                            })
-                            .collect::<Vec<u32>>()
-                    } else {
-                        v.into_u32().collect::<Vec<u32>>()
-                    }
-                }),
+                indices: reader.read_indices().context("GLTF mesh must contain an index buffer")?.into_u32().collect_vec(),
```
Comment (ten3roberts) [inline crates/model_import/src/gltf/mod.rs:54]: Wont this unconditionally flip the indices/winding?

File: crates/std/src/uncategorized/mesh.rs [hunk 2]
```diff
@@ -113,22 +101,18 @@ impl Mesh {
...
     }
     pub fn apply_skin(&mut self, joint_matrices: &[Mat4]) {
-        if let (Some(positions), Some(weights), Some(indices)) = (
-            &mut self.positions,
-            &self.joint_weights,
-            &self.joint_indices,
-        ) {
-            for ((position, weight), index) in
-                positions.iter_mut().zip(weights.iter()).zip(indices.iter())
+        if let (Some(weights), Some(indices)) = (&self.joint_weights, &self.joint_indices) {
```
Comment (ten3roberts) [inline crates/std/src/uncategorized/mesh.rs:110]: We could use `zip_eq` here if we expect them to be of the same length.

If not, we could use `izip!`to not have nested tuples

File: libs/unity_parser/src/asset.rs [hunk 3]
```diff
@@ -33,8 +34,12 @@ impl Asset {
             BlendWeight,
             BlendIndices,
         }
-        let vertex_count = asset["Mesh"]["m_VertexData"]["m_VertexCount"].as_i64().unwrap() as usize;
-        let vertex_data_size = asset["Mesh"]["m_VertexData"]["m_DataSize"].as_i64().unwrap() as usize;
+        let vertex_count = asset["Mesh"]["m_VertexData"]["m_VertexCount"]
```
Comment (ten3roberts) [inline libs/unity_parser/src/asset.rs:37]: Since we are now returning a result we could make the `[]` indexing fallible

**Comment by ten3roberts** [comment]
I left some comments.

Apart from that it looks like a good change and is something that I started doing; namely making sure that if vertex uvs exist the position will exist as well.

Having `Option<Vec<_>>` is usally a tiny bit of code smell and implies there is a better solution since the `Vec` is already a functor, I.e; an empty Vec has the same meaning as None, and the difference between them is very subtle and if that difference is important it could be expressed in a different way

File: crates/model_import/src/gltf/mod.rs (hunk 1 above)
Comment (VinhTruongAmbient) [reply crates/model_import/src/gltf/mod.rs:54]: Yep, this was the case before as well, as `flip_indices` was a constant `true`.

Comment (ten3roberts) [reply crates/model_import/src/gltf/mod.rs:54]: Oh I see now. Welp, if we need that behavior we have the history

File: crates/std/src/uncategorized/mesh.rs (hunk 2 above)
Comment (VinhTruongAmbient) [reply crates/std/src/uncategorized/mesh.rs:110]: Yeah sure. I will have another look on a follow-up pass.

**Comment by VinhTruongAmbient** [comment]
> I left some comments.
> 
> Apart from that it looks like a good change and is something that I started doing; namely making sure that if vertex uvs exist the position will exist as well.
> 
> Having `Option<Vec<_>>` is usally a tiny bit of code smell and implies there is a better solution since the `Vec` is already a functor, I.e; an empty Vec has the same meaning as None, and the difference between them is very subtle and if that difference is important it could be expressed in a different way

Thanks. I agree with making sure that the mesh is actually valid (all vertex attributes have equal number of elements etc). Might do it on a separate PR.

File: crates/model_import/src/gltf/mod.rs (hunk 1 above)
Comment (VinhTruongAmbient) [reply crates/model_import/src/gltf/mod.rs:54]: Yeah. If we wanted to do it properly, here's what the GLTF 2.0 spec says:

> When a mesh primitive uses any triangle-based topology (i.e., triangles, triangle strip, or triangle fan), the determinant of the node’s global transform defines the winding order of that primitive. If the determinant is a positive value, the winding order triangle faces is counterclockwise; in the opposite case, the winding order is clockwise.

**Pull request merged**